### DISCUSSION TITLE
Welcome to DACN Discussions!

### DISCUSSION THREAD

**Discussion by laugh12321** [laugh12321/Hyperspectral-Imagery-Unmixing#5] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ShallowDm** [comment]
tensorflow是要用2.3吗，为什么文档里面要求>=2.4.4

**Reply by laugh12321** [reply]
2.4.4应该也是可以的，README没有跟着更新，你觉得不保险的话就用2.3

**Reply by ShallowDm** [reply]
感谢回复，我目前用的2.6的tensorflow，运行没有模块缺失的错误了，但是运行到_get_model函数的时候，里面的f值为'__builtins__'导致ast.literal_eval函数识别报错，里面的node_or_string参数值为"<_ast.Name object at 0x0000013CADA42DF0>"，这个问题有遇到过吗

**Reply by laugh12321** [reply]
建议你不要用`ast.literal_eval`, 将所有的`ast.literal_eval`替换为`eval`。如果还有别的BUG, 你可以使用这个提交版本👉[0314bb4ef8f22ab85e8e739f02ccd86564af4d88](https://github.com/laugh12321/DACN/tree/0314bb4ef8f22ab85e8e739f02ccd86564af4d88)

**Reply by ShallowDm** [reply]
好的，我试试

**Reply by ShallowDm** [reply]
![运行](https://user-images.githubusercontent.com/83915800/234465583-5dd27d42-61ad-4ce4-a4fb-7c52544ba717.png)
这个运行效果正常吗

**Reply by ShallowDm** [reply]
训练到139次时出现这个
![image](https://user-images.githubusercontent.com/83915800/234546383-06e09df6-7a7a-4d7c-ba5a-7365d8a694a6.png)
然后开始重新训练了，而且进度条比之前多了几百，这个是正常现象吗
![image](https://user-images.githubusercontent.com/83915800/234546895-6a91785a-d4ca-4a8c-8ea0-b40e41683fa3.png)

**Reply by laugh12321** [reply]
所有的警告都可以不用管，他的意思是这种方法已经弃用，建议用新方法。你觉得不美观可以用warning.ignore。所谓的重新训练只是打印出警告后进度条另起一行重新刷新吧。

**Reply by ShallowDm** [reply]
好的，目前我在修改过eval函数后已经在kaggle上运行结束，
![image](https://user-images.githubusercontent.com/83915800/236159999-cdcd5fb9-f383-4022-89c6-3752b57ac6cd.png)
这个最终的解混数据会直接输出到哪个文件夹吗

**Reply by laugh12321** [reply]
最终结果以及模型会保存到`results`文件夹

**Reply by ShallowDm** [reply]
最后输出的应该时端元种类和丰度吧，我看好像文件夹里面好像没有符合条件的，可能是线上运行导致输出结果没有保存到正确的地址，这个代码的具体输出结果是保存excel里面的吗？

**Reply by ShallowDm** [reply]
还有想问一下那个图片文件夹里面的图片是运行后自动生成的吗？还是你根据数据自己做的，谢谢

**Reply by laugh12321** [reply]
图片当然是自己画的，不提供代码。程序只生成预测数据指标

**Reply by laugh12321** [reply]
> 最后输出的应该时端元种类和丰度吧，我看好像文件夹里面好像没有符合条件的，可能是线上运行导致输出结果没有保存到正确的地址，这个代码的具体输出结果是保存excel里面的吗？

你说的是Fig 5吧。那个是我根据训练完的模型预测结果然后画出来的。程序训练完只给指标数据和一个模型

**Reply by ShallowDm** [reply]
也就是说程序输出的是四种深度模型和四种模型对两组高光谱图像解混的预测结果，这样理解正确吗？还是需要使用另外的代码利用训练好的模型进行解混，另外fig5是图片吗

**Reply by laugh12321** [reply]
1. 首先程序运行完后，所有结果会存放在`results`文件夹下;
2. 其次，在`results`文件夹下共有8个子文件，分别对应4种方法、2种数据;
3. 其中，在每个子文件夹下都包含根据当前所有的`experiment_{num}`汇总计算得到的评估报告`report.csv`;
4. 在每个`experiment_{num}`下都包含了当前的评估指标`inference_metrics.csv`、最优模型`model.h5`、训练指标历史`training_metrics.csv`以及一张预测结果图`pred.npy`;
5. 其中，`Fig 2 3`是根据`training_metrics.csv`绘制出，`Fig 4`是根据`report.csv`绘制的，而`Fig 5`则是根据`pred.npy`绘制。
6. 所有数据均真是可靠，与README所绘制结果一致。

**Reply by ShallowDm** [reply]
原来如此，感谢解释

**Reply by laugh12321** [reply]
这些在源码中都有体现

**Reply by ShallowDm** [reply]
截取了前160个像素得到的丰度表示，
![image](https://github.com/laugh12321/DACN/assets/83915800/a2754369-07c6-4161-ab04-f69e7f97ba3d)

**Reply by laugh12321** [reply]
我没有用过这种方式表示丰度，是直接展示可视化结果。如果你想评价每种端元的好坏，`experiment_{num}/inference_metrics.csv`中已经给出了不同端元的RMSE。

**Reply by ShallowDm** [reply]
这个预测结果图文件不是直接生成的吗

**Comment by shayerfg** [comment]
in the result file i did not get any file named pred.npy

**Reply by ShallowDm** [reply]
result里面储存了四种模型的保存结果，里面的experiment文件夹里面有pred.npy文件

**Reply by shayerfg** [reply]
<img width="255" alt="Capture d’écran 2023-05-23 à 13 29 59" src="https://github.com/laugh12321/DACN/assets/129570421/b2563359-6565-4845-8c00-128341c536b0">
 i have no pred.npy in the experiment

**Reply by ShallowDm** [reply]
有没有可能下错了版本，github库里面有的，可以直接在github结果文件里面下载试试

**Reply by shayerfg** [reply]
i downloaded the whole file  and i tried  executing just the cnn_pixel_based  but i didnt get any file named pred.npy also im having an error in the time_metrics.py .I have an error as if I haven't downloaded Keras, even though I have downloaded it, and even with this error, the code is still executing. Do you think this is normal?
<img width="1023" alt="Capture d’écran 2023-05-23 à 13 42 54" src="https://github.com/laugh12321/DACN/assets/129570421/0f7897e2-4a17-458e-935b-e39e8c97ff6d">

**Reply by ShallowDm** [reply]
应该是python软件包的问题，但我是直接不能运行还报错，你这种情况没见过，可以尝试重新下载tensorflow和keras，并且这两个软件包的版本需要对应，网上有对应的版本对应表

**Reply by shayerfg** [reply]
i did try re install keras and tensorflow but it didn’t work , can i please send you the file i have that contains the whole code and you try to executes it to see if you get the pred.npy file?

**Reply by ShallowDm** [reply]
额，我是在线上运行的程序，你可以使用kaggle网站线上跑模型，它的训练速度也很快，只需要将整个代码的文件夹打包上传网站就行了

**Reply by shayerfg** [reply]
oh thank you then im going to try and run it on the website of kaggle and see

**Reply by ShallowDm** [reply]
没事，我现在也在学习这个代码的框架

**Reply by shayerfg** [reply]
that's great. i have one last question please how did you get the Ground-truth and estimated abundances obtained for each endmember material in the Urban datasets

**Reply by ShallowDm** [reply]
我个人觉得是pred.npy文件里面的数组，里面的数组数量刚好和官方给的最终成员的个数相同，我认为是它将每个像素的估计丰度储存在了这个由数组构成的数据文件中

**Reply by shayerfg** [reply]
Thank you so much for this precious information

**Reply by shayerfg** [reply]
please can you send me the. file that you are working on because i tried to execute it on kaggle and didn’t get the pred.py file  .

**Reply by ShallowDm** [reply]
我重新执行了一遍，也没有生成这个pred文件，至于这个文件你可以直接在github上下载

**Reply by laugh12321** [reply]
要生成`pred.npy`需要修改`src/model/evaluate_unmixing.py`代码, 让它在预测时将第一个预测结果保存为`pred.npy`.

**Reply by shayerfg** [reply]
i did modify the code but the file i got have 226 Ko meanwhile your file have 2,3 Mo  here's what i did modify in the code 
<img width="1011" alt="Capture d’écran 2023-05-26 à 13 43 39" src="https://github.com/laugh12321/DACN/assets/129570421/21521d1e-f4b3-4b18-ae7e-d23ddaf3339d">
<img width="1011" alt="Capture d’écran 2023-05-26 à 13 44 13" src="https://github.com/laugh12321/DACN/assets/129570421/9e353254-8947-4c20-ac16-3c0f2cb13ad7">

**Reply by shayerfg** [reply]
it is not saving everything i guess . that's why i couldn’t get the abundance map

**Reply by shayerfg** [reply]
for the urban dataset im having (9425, 6) instead of (94249, 6) how to fix it

**Reply by laugh12321** [reply]
`np.save(os.path.join(dest_path, "pred.npy"), y_pred)` 保存的是这一个batch的所有预测结果，而我在上文已经提到你只需要保一个预测结果即可。所以应该改成`np.save(os.path.join(dest_path, "pred.npy"), y_pred[0])`。

关于你说的`ubran`数据集尺寸不对的问题，由于时间太久远我也记得不是很清楚。`urban`的数据集真实尺寸应该是`(307, 307, 6)`,展开后应该是`(94249, 6)`，`6`代表它有6个端元。

> 如果没有记错的话`np.save(os.path.join(dest_path, "pred.npy"), y_pred[0])`保存的尺寸应该是`(94249, 6)`，你需要把它转成`(307, 307, 6)`，才能绘制端元丰度。

**Reply by shayerfg** [reply]
Unfortunately  even  np.save(os.path.join(dest_path, "pred.npy"), y_pred[0]) did not save it as (94249, 6)   it saves it as (6, 6).   if you have  the file that saves it  can you share it please?

**Reply by laugh12321** [reply]
抱歉，因为这个项目太早了，有些回复并没有经过仔细推敲。因为在`evaluate_unmixing.py`中它是一次预测一个batch，所以我们只需要batch中任意一张图的预测结果即可，相比较而言第一张比较容易拿到。但可能获取的方式可能并不是`y_pred[0]`，也许是`y_pred[:,0]`或者`y_pred[:,:,0]`你可以按照这个思路尝试一下。

**Reply by shayerfg** [reply]
Iit did not work either . for the  jasper dataset  im having (1000, 4) instead of (10000,4)   i also tried y_pred[:,:,0] did not work

**Reply by shayerfg** [reply]
it may be that the y_pred does not have all the predections?

**Reply by shayerfg** [reply]
I think the problem is in the y_pred it does not have all the predictions

**Reply by laugh12321** [reply]
有空我帮你解决一下，这段时间太忙了

**Reply by shayerfg** [reply]
I need it before this Friday if you had a free time before Friday i will really appreciate it
